src/ppom-storage.ts

```diff
@@ -0,0 +1,206 @@
+/**
+ * FileInfo Type
+ * Defined type for information about file saved in storage backend.
+ */
+type FileInfo = {
+  name: string;
+  chainId: string;
+  version: string;
+  checksum: string;
+};
+
+/**
+ * PPOMFileMetadata Type
+ * This is type of metadata about files saved in storage,
+ * this information is saved in PPOMController state.
+ */
+export type PPOMFileMetadata = FileInfo[];
+
+/**
+ * StorageKey Type
+ * This defines type of key that is used for indexing file data saved in StorageBackend.
+ */
+export type StorageKey = {
+  name: string;
+  chainId: string;
+};
+
+/**
+ * StorageBackend Type
+ * This defines type for storage backend implementation.
+ * There will be different storage implementations depending on platform:
+ * 1. extension - indexDB
+ * 2. mobile app - <TBD>
+ */
+export type StorageBackend = {
+  read(key: StorageKey, checksum: string): Promise<ArrayBuffer>;
+  write(key: StorageKey, data: ArrayBuffer, checksum: string): Promise<void>;
+  delete(key: StorageKey): Promise<void>;
+  dir(): Promise<StorageKey[]>;
+};
+
+/**
+ * PPOMStorage class
+ * This class is responsible for managing the local storage
+ * It provides the following functionalities:
+ * 1. Sync the metadata with the version info from the cdn
+ * 2. Read a file from the local storage
+ * 3. Write a file to the local storage
+ *
+ * It also validates the checksum of the file when reading and writing in order to
+ * detect corrupted files or files that are not up to date
+ */
+export class PPOMStorage {
+  readonly #storageBackend: StorageBackend;
+
+  readonly #readMetadata: () => PPOMFileMetadata;
```

Comment:
maybe rename `readMetadata` to `readStorageMetadata` based on https://github.com/MetaMask/ppom-validator/blob/b64259cb8e9998d187b50f62c814941cef33e692/src/ppom-controller.ts#LL143C1-L151C8

```
    this.#storage = new PPOMStorage({
      storageBackend,
      readMetadata: () => {
        return this.state.storageMetadata;
      },
      writeMetadata: (metadata) => {
        this.update({ storageMetadata: metadata });
      },
    });
```
 to preserve the name of the property we are reading and writing to. This is also to help separate the meaning between this method and `readFile`